File: backend/app/schemas/custom_field.py
import uuid
from datetime import datetime
from typing import Any, Literal, Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SelectOption(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    label: str = Field(min_length=1, max_length=100)
    color: str = Field(min_length=4, max_length=7)


# ── Definition Schemas ──


class CustomFieldDefinitionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    field_type: Literal[
        "text", "number", "select", "multi_select",
        "date", "checkbox", "url", "person"
    ]
    description: str | None = None
    options: list[SelectOption] | None = None
    is_required: bool = False

    @model_validator(mode="after")
    def validate_options(self) -> Self:
        if self.field_type in ("select", "multi_select"):
            if not self.options or len(self.options) == 0:
                raise ValueError(
                    "options required for select/multi_select fields"
                )
        elif self.options:
            self.options = None
        return self


class CustomFieldDefinitionUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    options: list[SelectOption] | None = None
    is_required: bool | None = None


class CustomFieldDefinitionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    board_id: UUID
    name: str
    field_type: str
    description: str | None = None
    options: list[SelectOption] | None = None
    is_required: bool
    position: float
    created_at: datetime
    updated_at: datetime


class CustomFieldReorder(BaseModel):
    field_ids: list[UUID]


# ── Value Schemas ──


class CustomFieldValueSet(BaseModel):
    field_definition_id: UUID
    value_text: str | None = None
    value_number: float | None = None
    value_json: Any | None = None
    value_date: datetime | None = None


class BulkFieldValueSet(BaseModel):
    values: list[CustomFieldValueSet]


class CustomFieldValueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    field_definition_id: UUID
    value_text: str | None = None
    value_number: float | None = None
    value_json: Any | None = None
    value_date: datetime | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/custom-field.ts
export type CustomFieldType =
  | 'text'
  | 'number'
  | 'select'
  | 'multi_select'
  | 'date'
  | 'checkbox'
  | 'url'
  | 'person'

export interface SelectOption {
  id: string
  label: string
  color: string
}

export interface CustomFieldDefinition {
  id: string
  board_id: string
  name: string
  field_type: CustomFieldType
  description: string | null
  options: SelectOption[] | null
  is_required: boolean
  position: number
  created_at: string
  updated_at: string
}

export interface CustomFieldDefinitionCreate {
  name: string
  field_type: CustomFieldType
  description?: string
  options?: SelectOption[]
  is_required?: boolean
}

export interface CustomFieldDefinitionUpdate {
  name?: string
  description?: string
  options?: SelectOption[]
  is_required?: boolean
}

export interface CustomFieldValue {
  id: string
  field_definition_id: string
  value_text: string | null
  value_number: number | null
  value_json: unknown | null
  value_date: string | null
  created_at: string
  updated_at: string
}

export interface CustomFieldValueSet {
  field_definition_id: string
  value_text?: string | null
  value_number?: number | null
  value_json?: unknown | null
  value_date?: string | null
}

export interface FieldTypeConfig {
  type: CustomFieldType
  label: string
  icon: string
  description: string
}

export const FIELD_TYPE_CONFIGS: FieldTypeConfig[] = [
  { type: 'text',         label: 'Text',         icon: 'Type',          description: 'Short or long text' },
  { type: 'number',       label: 'Number',       icon: 'Hash',          description: 'Numeric value' },
  { type: 'select',       label: 'Select',       icon: 'ChevronDown',   description: 'Single choice from options' },
  { type: 'multi_select', label: 'Multi-Select', icon: 'ListChecks',    description: 'Multiple choices from options' },
  { type: 'date',         label: 'Date',         icon: 'Calendar',      description: 'Date picker' },
  { type: 'checkbox',     label: 'Checkbox',     icon: 'CheckSquare',   description: 'True/false toggle' },
  { type: 'url',          label: 'URL',          icon: 'Link',          description: 'Web address' },
  { type: 'person',       label: 'Person',       icon: 'UserCircle',    description: 'Team member or agent' },
]
